NPE in CtTypeReferenceImpl#isInSamePackage in noclasspath

Version 7.2.0 and 7.3.0-SNAPSHOT

I receive this exception when I try to run Spoon in noclasspath:

```java
Exception in thread "main" spoon.SpoonException: Printing of spoon.support.reflect.code.CtThisAccessImpl on path 
	CtClassImpl:InCondition
	CtConstructorImpl:<init>
	CtBlockImpl
	CtAssignmentImpl
	CtFieldWriteImpl
	CtThisAccessImpl
at position (/home/tdurieux/git/patch-clustering/Expression.java:348) failed
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:354)
	at add.features.detector.repairpatterns.CopyPasteDetector$1.scan(CopyPasteDetector.java:67)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.printCtFieldAccess(DefaultJavaPrettyPrinter.java:871)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtFieldWrite(DefaultJavaPrettyPrinter.java:804)
	at spoon.support.reflect.code.CtFieldWriteImpl.accept(CtFieldWriteImpl.java:27)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:345)
	at add.features.detector.repairpatterns.CopyPasteDetector$1.scan(CopyPasteDetector.java:59)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtAssignment(DefaultJavaPrettyPrinter.java:494)
	at spoon.support.reflect.code.CtAssignmentImpl.accept(CtAssignmentImpl.java:55)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:345)
	at add.features.detector.repairpatterns.CopyPasteDetector$1.scan(CopyPasteDetector.java:67)
	at spoon.reflect.visitor.ElementPrinterHelper.writeStatement(ElementPrinterHelper.java:182)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtBlock(DefaultJavaPrettyPrinter.java:526)
	at spoon.support.reflect.code.CtBlockImpl.accept(CtBlockImpl.java:48)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:345)
	at add.features.detector.repairpatterns.CopyPasteDetector$1.scan(CopyPasteDetector.java:67)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtConstructor(DefaultJavaPrettyPrinter.java:704)
	at spoon.support.reflect.declaration.CtConstructorImpl.accept(CtConstructorImpl.java:57)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:345)
	at add.features.detector.repairpatterns.CopyPasteDetector$1.scan(CopyPasteDetector.java:67)
	at spoon.reflect.visitor.ElementPrinterHelper.writeElementList(ElementPrinterHelper.java:196)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtClass(DefaultJavaPrettyPrinter.java:622)
	at spoon.support.reflect.declaration.CtClassImpl.accept(CtClassImpl.java:68)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:345)
	at add.features.detector.repairpatterns.CopyPasteDetector$1.scan(CopyPasteDetector.java:67)
	at add.features.detector.repairpatterns.CopyPasteDetector.detect(CopyPasteDetector.java:75)
	at add.features.detector.repairpatterns.RepairPatternDetector.analyze(RepairPatternDetector.java:45)
	at add.features.detector.repairpatterns.RepairPatternDetector.analyze(RepairPatternDetector.java:15)
	at add.main.Launcher.execute(Launcher.java:154)
	at add.main.Launcher.main(Launcher.java:167)
Caused by: java.lang.NullPointerException
	at spoon.support.reflect.reference.CtTypeReferenceImpl.isInSamePackage(CtTypeReferenceImpl.java:665)
	at spoon.support.reflect.reference.CtTypeReferenceImpl.canAccess(CtTypeReferenceImpl.java:656)
	at spoon.support.reflect.reference.CtTypeReferenceImpl.getAccessType(CtTypeReferenceImpl.java:691)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtTypeReference(DefaultJavaPrettyPrinter.java:1824)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtTypeReferenceWithoutGenerics(DefaultJavaPrettyPrinter.java:1808)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.visitCtThisAccess(DefaultJavaPrettyPrinter.java:948)
	at spoon.support.reflect.code.CtThisAccessImpl.accept(CtThisAccessImpl.java:28)
	at spoon.reflect.visitor.DefaultJavaPrettyPrinter.scan(DefaultJavaPrettyPrinter.java:345)
	... 29 more
```

--- a/src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java
+++ b/src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java
@@ -574,6 +574,10 @@ public class CtTypeReferenceImpl<T> extends CtReferenceImpl implements CtTypeRef
 	@Override
 	public boolean canAccess(CtTypeReference<?> type) {
 		try {
+			if (type == null) {
+				//noclasspath mode
+				return true;
+			}
 			if (type.getTypeDeclaration() == null) {
 				return true;
 			}
@@ -664,7 +668,12 @@ public class CtTypeReferenceImpl<T> extends CtReferenceImpl implements CtTypeRef
 	}
 
 	private boolean isInSamePackage(CtTypeReference<?> type) {
-		return type.getTopLevelType().getPackage().getSimpleName().equals(this.getTopLevelType().getPackage().getSimpleName());
+		CtPackageReference thisPackage = this.getTopLevelType().getPackage();
+		CtPackageReference otherPackage = type.getTopLevelType().getPackage();
+		if (thisPackage == null || otherPackage == null) {
+			return true;
+		}
+		return thisPackage.getQualifiedName().equals(otherPackage.getQualifiedName());
 	}
 
 	@Override

--- a/src/test/java/spoon/test/reference/TypeReferenceTest.java
+++ b/src/test/java/spoon/test/reference/TypeReferenceTest.java
@@ -21,6 +21,8 @@ import spoon.Launcher;
 import spoon.SpoonModelBuilder;
 import spoon.compiler.SpoonResource;
 import spoon.compiler.SpoonResourceHelper;
+import spoon.reflect.CtModel;
+import spoon.reflect.code.CtAssignment;
 import spoon.reflect.code.CtConstructorCall;
 import spoon.reflect.code.CtExpression;
 import spoon.reflect.code.CtFieldRead;
@@ -676,5 +678,27 @@ public class TypeReferenceTest {
 		typeRef.getPackage().setImplicit(false);
 		assertFalse(typeRef.isImplicitParent());
 		assertFalse(typeRef.getPackage().isImplicit());
+	}
+
+	@Test
+	public void testIsInTheSamePackageNoClasspath() {
+		// contract: we should not get NPE within canAccess() in noclasspath mode
+		final Launcher launcher = new Launcher();
+		launcher.addInputResource("./src/test/resources/noclasspath/A5.java");
+		launcher.getEnvironment().setNoClasspath(true);
+		CtModel model = launcher.buildModel();
+		CtType<?> type = model.getAllTypes().stream().findFirst().get();
+
+		List<CtLocalVariable> vars = type.getElements(new TypeFilter<>(CtLocalVariable.class));
+		List<CtField> fields = type.getElements(new TypeFilter<>(CtField.class));
+		List<CtInvocation> invocations = type.getElements(new TypeFilter<>(CtInvocation.class));
+
+		// test null type
+		CtTypeReference tr1 = invocations.get(1).getType();
+		assertTrue(fields.get(0).getReference().getType().canAccess(tr1));
+
+		// test type with null package
+		CtTypeReference tr2 = fields.get(0).getType();
+		assertTrue(vars.get(0).getReference().getType().canAccess(tr2));
 	}
 }

--- a/src/test/resources/noclasspath/A5.java
+++ b/src/test/resources/noclasspath/A5.java
@@ -1,0 +1,9 @@
+public class A5<T> extends UnknownKlass {
+
+    T a;
+
+    void m1() {
+        X x;
+        someMethod();
+    }
+}

